thorlcr/master/thgraphmanager.cpp

```diff
@@ -1463,63 +1464,146 @@ void thorMain(ILogMsgHandler *logHandler, const char *wuid, const char *graphNam
-                                w->setDebugValue(instance, "1", true);
-
-                            if (jobManager->queryExitException())
+                            else if (querySessionManager().sessionStopped(agentSessionID, 0))
                             {
-                                // NB: exitException has already been relayed.
-                                jobManager->clearExitException();
+                                WARNLOG("Discarding agentless job: wuid=%s, graph=%s", currentWuid.str(), currentGraphName.str());
+                                currentWuid.clear();
                             }
                             else
                             {
-                                switch (w->getState())
+                                SCMStringBuffer jobVersion;
+                                workunit->getDebugValue("platformVersion", jobVersion);
+                                bool mismatch = false;
+                                const char *imagePattern = getenv("imagePattern");
+                                const char *imageVersion = nullptr;
+                                if (isEmptyString(imagePattern))
+                                {
+                                    IWARNLOG("imagePattern not set in environment");
+                                    imageVersion = "unknown";
+                                    runtimePlatformVersion.set(imageVersion);
+                                }
+                                else
+                                {
+                                    imageVersion = strrchr(imagePattern, ':');
+                                    if (!imageVersion)
+                                    {
+                                        IWARNLOG("imagePattern has unexpected format: %s", imagePattern);
+                                        imageVersion = "unknown";
+                                        runtimePlatformVersion.set(imageVersion);
+                                    }
+                                    else
+                                    {
+                                        imageVersion++;
+                                        if (jobVersion.length())
+                                        {
+                                            if (0 == runtimePlatformVersion.length()) // 1st time
+                                                runtimePlatformVersion.set(jobVersion.str());
+                                            else if (!streq(jobVersion.str(), runtimePlatformVersion.str()))
+                                                mismatch = true;
+                                        }
+                                        else if (runtimePlatformVersion.length())
+                                        {
+                                            if (!streq(imageVersion, runtimePlatformVersion.str())) // can only happen if a previous job had launched instance with a jobVersion
```

Comment:
Can also occur if imageVersion is "unknown".  That should also match if no jobVersion is specified.

Reply:
as it stands, it can't get here if "unknown". "unknown" only set if didn't reach here.